File: backend/app/schemas/transport.py
import uuid
from datetime import time
from typing import Optional

from pydantic import BaseModel, Field


class TransportVehicleOut(BaseModel):
    id: uuid.UUID
    school_id: uuid.UUID
    name: str
    registration_no: Optional[str]
    capacity: int
    driver_name: Optional[str]
    status: str


class TransportVehicleCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    registration_no: Optional[str] = Field(default=None, max_length=64)
    capacity: int = Field(default=0, ge=0, le=100000)
    driver_name: Optional[str] = Field(default=None, max_length=150)
    status: str = Field(default="active", max_length=32)


class TransportVehicleUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=120)
    registration_no: Optional[str] = Field(default=None, max_length=64)
    capacity: Optional[int] = Field(default=None, ge=0, le=100000)
    driver_name: Optional[str] = Field(default=None, max_length=150)
    status: Optional[str] = Field(default=None, max_length=32)


class TransportRouteOut(BaseModel):
    id: uuid.UUID
    school_id: uuid.UUID
    name: str
    code: Optional[str]
    description: Optional[str]
    is_active: bool


class TransportRouteCreate(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    code: Optional[str] = Field(default=None, max_length=50)
    description: Optional[str] = Field(default=None, max_length=500)
    is_active: bool = True


class TransportRouteUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=150)
    code: Optional[str] = Field(default=None, max_length=50)
    description: Optional[str] = Field(default=None, max_length=500)
    is_active: Optional[bool] = None


class TransportRouteStopOut(BaseModel):
    id: uuid.UUID
    route_id: uuid.UUID
    name: str
    sequence: int
    pickup_time: Optional[time]
    drop_time: Optional[time]


class TransportRouteStopCreate(BaseModel):
    route_id: uuid.UUID
    name: str = Field(min_length=1, max_length=150)
    sequence: int = Field(default=1, ge=1, le=100000)
    pickup_time: Optional[time] = None
    drop_time: Optional[time] = None


class TransportRouteStopUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=150)
    sequence: Optional[int] = Field(default=None, ge=1, le=100000)
    pickup_time: Optional[time] = None
    drop_time: Optional[time] = None


class StudentTransportAssignmentOut(BaseModel):
    id: uuid.UUID
    school_id: uuid.UUID
    student_id: uuid.UUID
    route_id: uuid.UUID
    stop_id: Optional[uuid.UUID]
    vehicle_id: Optional[uuid.UUID]
    status: str


class AssignStudentTransportRequest(BaseModel):
    student_id: uuid.UUID
    route_id: uuid.UUID
    vehicle_id: Optional[uuid.UUID] = None
    stop_id: Optional[uuid.UUID] = None
    status: str = Field(default="active", max_length=32)


class UpdateStudentTransportRequest(BaseModel):
    route_id: Optional[uuid.UUID] = None
    vehicle_id: Optional[uuid.UUID] = None
    stop_id: Optional[uuid.UUID] = None
    status: Optional[str] = Field(default=None, max_length=32)

File: frontend/src/api/transport.ts
import { safeRequest } from "./httpSafe";

export type TransportVehicle = {
  id: string;
  school_id: string;
  name: string;
  registration_no: string | null;
  capacity: number;
  driver_name: string | null;
  status: string;
};

export type TransportRoute = {
  id: string;
  school_id: string;
  name: string;
  code: string | null;
  description: string | null;
  is_active: boolean;
};

export type TransportRouteStop = {
  id: string;
  route_id: string;
  name: string;
  sequence: number;
  pickup_time: string | null;
  drop_time: string | null;
};

export type StudentTransportAssignment = {
  id: string;
  school_id: string;
  student_id: string;
  route_id: string;
  stop_id: string | null;
  vehicle_id: string | null;
  status: string;
};

export async function getTransportVehicles(params?: { status?: string }): Promise<TransportVehicle[]> {
  const r = await safeRequest<TransportVehicle[]>({
    method: "get",
    url: "/transport/vehicles",
    params,
  });
  if (!r.ok) throw new Error(String(r.status));
  return r.data;
}

export async function createTransportVehicle(payload: {
  name: string;
  registration_no?: string | null;
  capacity: number;
  driver_name?: string | null;
  status: string;
}): Promise<TransportVehicle> {
  const r = await safeRequest<TransportVehicle>({
    method: "post",
    url: "/transport/vehicles",
    data: payload,
  });
  if (!r.ok) throw new Error(String(r.status));
  return r.data;
}

export async function getTransportRoutes(): Promise<TransportRoute[]> {
  const r = await safeRequest<TransportRoute[]>({ method: "get", url: "/transport/routes" });
  if (!r.ok) throw new Error(String(r.status));
  return r.data;
}

export async function createTransportRoute(payload: {
  name: string;
  code?: string | null;
  description?: string | null;
  is_active: boolean;
}): Promise<TransportRoute> {
  const r = await safeRequest<TransportRoute>({
    method: "post",
    url: "/transport/routes",
    data: payload,
  });
  if (!r.ok) throw new Error(String(r.status));
  return r.data;
}

export async function getRouteStops(routeId: string): Promise<TransportRouteStop[]> {
  const r = await safeRequest<TransportRouteStop[]>({
    method: "get",
    url: `/transport/routes/${routeId}/stops`,
  });
  if (!r.ok) throw new Error(String(r.status));
  return r.data;
}

export async function getTransportAssignments(params?: {
  route_id?: string;
  vehicle_id?: string;
}): Promise<StudentTransportAssignment[]> {
  const r = await safeRequest<StudentTransportAssignment[]>({
    method: "get",
    url: "/transport/student-assignments",
    params,
  });
  if (!r.ok) throw new Error(String(r.status));
  return r.data;
}

